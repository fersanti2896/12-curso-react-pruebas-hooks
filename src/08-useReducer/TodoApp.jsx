import { useTodos } from '../hooks';
import { TodoAdd, TodoList } from './components';


export const TodoApp = () => {
    const { handleNewTodo, handleDeleteTodo, handleToogleTodo, todos, todosCount, pendingTodosCount } = useTodos();
    
    return (
        <>
            <h2>Todos: { todosCount } | Pendientes: { pendingTodosCount }</h2>
            <hr/>

            <div className='row'>
                <div className='col-7'>
                    <TodoList todos={ todos } 
                              onDeleteTodo={ handleDeleteTodo }
                              onToogleTodo={ handleToogleTodo } />
                </div>

                <div className='col-5'>
                    <h4>Agregar Todo</h4>
                    <hr />

                    <TodoAdd onNewTodo={ handleNewTodo } />
                </div>   
            </div>
        </>
    )
}
