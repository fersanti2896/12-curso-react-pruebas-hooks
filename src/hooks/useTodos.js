import { useEffect, useReducer, useState } from 'react';
import { todoReducer } from '../08-useReducer/todoReducer';

/**
 * Funci贸n que se encarga de cargar los todos.
 * @returns Devuelve un arreglo de JSON de los todos
 */
const init = () => {
    return JSON.parse( localStorage.getItem('todos') ) || [];
}

export const useTodos = () => {
    const [ todos, dispatch ] = useReducer( todoReducer, [], init );
    
    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify(todos) );
    }, [ todos ])
    
    /**
     * Funci贸n que agrega un todo al arreglo de todos.
     * @param {Object} todo Ejemplo: { id: 1, description: 'Gema del porder', done: false }
     */
    const handleNewTodo = ( todo ) => {
        const action = {
            type: '[TODO] Add Todo',
            payload: todo
        }
    
        dispatch( action );
    }

    /**
     * Funci贸n que elimina un todo por su id.
     * @param {string} id 
     */
    const handleDeleteTodo = ( id ) => {
        dispatch({
            type: '[TODO] Remove Todo',
            payload: id
        });
    }

    /**
     * Funci贸n que marca un todo como completado por su id.
     * @param {string} id 
     */
    const handleToogleTodo = ( id ) => {
        dispatch({
            type: '[TODO] Toggle Todo',
            payload: id
        })
    }

    return {
        handleDeleteTodo, 
        handleNewTodo, 
        handleToogleTodo,
        todos, 
        todosCount: todos.length,
        pendingTodosCount: todos.filter( todo => !todo.done ).length
    }
}
