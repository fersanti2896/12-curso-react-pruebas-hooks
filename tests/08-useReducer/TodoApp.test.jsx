import { render, screen } from '@testing-library/react';
import { TodoApp } from '../../src/08-useReducer/TodoApp';
import { useTodos } from '../../src/hooks/useTodos';

jest.mock('../../src/hooks/useTodos');

describe('Pruebas en <TodoApp />', () => { 
    useTodos.mockReturnValue({
        handleDeleteTodo: jest.fn(), 
        handleNewTodo: jest.fn(), 
        handleToogleTodo: jest.fn(), 
        pendingTodosCount: 1,
        todos: [
            { id: 1, description: 'Todo #1', done: false },
            { id: 2, description: 'Todo #2', done: true }
        ], 
        todosCount: 2
    });

    test('Debe hacer match con en snapshot.', () => { 
        const { container } = render( <TodoApp /> );

        expect( container ).toMatchSnapshot();
    });

    test('Debe de mostrar el componente correctamente.', () => { 
        render( <TodoApp /> );

        expect( screen.getByText('Todo #1') ).toBeTruthy();
        expect( screen.getByText('Todo #2') ).toBeTruthy();
        expect( screen.getByRole('textbox') ).toBeTruthy();
        expect( screen.getByRole('textbox').placeholder ).toBe('¿Qué hay que hacer?');
    });
});