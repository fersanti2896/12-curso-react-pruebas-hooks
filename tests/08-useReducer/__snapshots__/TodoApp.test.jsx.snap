// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp /> Debe hacer match con en snapshot. 1`] = `
<div>
  <h2>
    Todos: 
    2
     | Pendientes: 
    1
  </h2>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-7"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center "
          >
            Todo #1
          </span>
          <button
            class="btn btn-outline-danger"
          >
            Borrar
          </button>
        </li>
        <li
          class="list-group-item d-flex justify-content-between"
        >
          <span
            aria-label="span"
            class="align-self-center text-decoration-line-through"
          >
            Todo #2
          </span>
          <button
            class="btn btn-outline-danger"
          >
            Borrar
          </button>
        </li>
      </ul>
    </div>
    <div
      class="col-5"
    >
      <h4>
        Agregar Todo
      </h4>
      <hr />
      <form>
        <input
          class="form-control"
          name="description"
          placeholder="¿Qué hay que hacer?"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-success mt-2"
          type="submit"
        >
          Agregar
        </button>
      </form>
    </div>
  </div>
</div>
`;
